### PR TITLE
Basic Prover instructions implementation

### PR DESCRIPTION
**Pull request by spartucus**
This PR did:
* ADD, MUL, EQ, MOV, JMP, CJMP, MLOAD, MSTORE instructions are basically implemented.
* Some code format.

Need to be done:
* CALL, RET instruction constraints.
* Context constraints like CLK, PC.
* Memory constraint.
* Program constraint.

### CODE REVIEW

File: prover/src/arithmetic/add.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    lv: &[ExtensionTarget<D>; NUM_INST_COLS],
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    // Get ADD data from trace.
+    let is_add = lv[COL_S_ADD];
+    let output = lv[COL_OP_0];
+    let input0 = lv[COL_OP_1];
+    let input1 = lv[COL_OP_2];
+
+    let unreduced_output = builder.add_extension(input0, input1);
```
Comment (JiangXb-son) [inline prover/src/arithmetic/add.rs:49]: Why reuse the circuit of plonky2?

Comment (JiangXb-son) [reply prover/src/arithmetic/add.rs:49]: for u32_add, the trace should be
a + b*1 = d
d_lo + d_hi *2^32 = d
rangecheck(d_lo)
bool(d_hi)

Comment (JiangXb-son) [reply prover/src/arithmetic/add.rs:49]: New version:
for u32_add, the trace should be
a + b * 1 = d
d_lo + d_hi *2^32 = d
d_lo_lo16 + d_lo_hi_16 *2^16 = d_lo
range check(d_lo_lo16, d_lo_hi16)
bool(d_hi)

File: prover/src/builtins/range_check_u32.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    | INPUT | LIMB0 | LIMB1 | ... | LIMB15 |
+    +-------+-------+------+-------+-------+
+     **/
+
+    const BASE: usize = 4;
+
+    const COL_INPUT: usize = 0;
+    const COL_START_LIMBS: usize = Self::COL_INPUT + 1;
+
+    fn range_limbs(&self) -> Range<usize> {
```
Comment (JiangXb-son) [inline prover/src/builtins/range_check_u32.rs:33]: Use lookup argument pls

**Review by JiangXb-son — CHANGES_REQUESTED** [review]

File: prover/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    assert!(reg_index < REGISTER_NUM as u8);
+                    $step.regs[reg_index as usize]
+                }
+            };
+
+            $row[COL_OP_2] = F::from_canonical_u64(op2.0);
+        };
+    }
+
+    let trace: Vec<[F; NUM_INST_COLS]> = steps
```
Comment (JiangXb-son) [inline prover/src/utils.rs:26]: trace should be vec<vec<>>?

File: prover/src/flow/flow_stark.rs [hunk 4]
```diff
@@ -41,7 +43,12 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for ArithmeticSta
         P: PackedField<Scalar = FE>,
     {
         let lv = vars.local_values;
+        let nv = vars.next_values;
         mov::eval_packed_generic(lv, yield_constr);
+        jmp::eval_packed_generic(lv, yield_constr);
+
+        // every setp, clk increase 1.
+        yield_constr.constraint(nv[COL_CLK] * (nv[COL_CLK] - lv[COL_CLK] - P::ONES));
```
Comment (Softcloud88) [inline prover/src/flow/flow_stark.rs:43]: Should put this constraint and jmp logic into a CpuStark.

**Pull request merged**